blockchain/validate_test.go

```diff
@@ -988,3 +988,282 @@ func TestCheckTicketExhaustion(t *testing.T) {
+	//
+	//   ... -> bsvh -> btmp0 -> ... -> btmp#
+	// -------------------------------------------------------------------------
+
+	outs = g.OldestCoinbaseOuts()
+	for i := uint16(0); i < coinbaseMaturity; i++ {
+		blockName := fmt.Sprintf("btmp%d", i)
+		g.NextBlock(blockName, &outs[0], outs[1:])
+		g.SaveTipCoinbaseOuts()
+		g.AcceptTipBlock()
+		outs = g.OldestCoinbaseOuts()
+	}
+
+	// -------------------------------------------------------------------------
+	// Create block that spends the regular and stake transactions created above
+	// with versions that will no longer be allowed for new transactions after
+	// the explicit version upgrades agend to ensure the utxos remain spendable
+	// prior to the activation of the agenda.
+	//
+	// Note that this block and the temp blocks above will be undone later so
+	// the same outputs can be spent after the explicit version upgrades agenda
+	// is active as well.
+	//
+	//   ... -> btmp# -> bspend0
+	// -------------------------------------------------------------------------
+
+	outs = g.OldestCoinbaseOuts()
+	g.NextBlock("bspend0", &outs[0], outs[1:], func(b *wire.MsgBlock) {
+		const regSpendTxIdx = 2
+		assertRegularBlockTxVer("bsvh", regSpendTxIdx, ^uint16(0))
+		bsvh := g.BlockByName("bsvh")
+		regSpend := chaingen.MakeSpendableOut(bsvh, regSpendTxIdx, 0)
+		regSpendTx := g.CreateSpendTx(&regSpend, lowFee)
+		b.AddTransaction(regSpendTx)
+
+		const stakeSpendTxIdx = 3
+		assertStakeBlockTxVer("bsvh", stakeSpendTxIdx, ^uint16(0))
+		stakeSpend := chaingen.MakeSpendableStakeOut(bsvh, stakeSpendTxIdx, 2)
+		stakeSpendTx := g.CreateSpendTx(&stakeSpend, lowFee)
+		b.AddTransaction(stakeSpendTx)
+	})
+	g.AcceptTipBlock()
+
+	// -------------------------------------------------------------------------
+	// Set the harness tip back to stake validation height, invalidate the
+	// blocks above so the underlying chain is also set back to that svh, and
+	// then activate the explicit version upgrades agenda.
+	// -------------------------------------------------------------------------
+
+	g.SetTip("bsvh")
+	g.InvalidateBlockAndExpectTip("btmp0", nil, "bsvh")
+	g.RestoreCoinbaseOutsSnapshot("postSVH")
+	g.AdvanceFromSVHToActiveAgendas(voteID)
+
+	// replaceVers is a munge function which modifies the provided block by
+	// replacing the block, stake, and vote versions with the explicit version
+	// deployment version.
+	replaceVers := func(b *wire.MsgBlock) {
+		chaingen.ReplaceBlockVersion(int32(deploymentVer))(b)
+		chaingen.ReplaceStakeVersion(deploymentVer)(b)
+		chaingen.ReplaceVoteVersions(deploymentVer)(b)
+	}
+
+	// -------------------------------------------------------------------------
+	// Create enough blocks to allow the stake output created at stake
+	// validation height to mature.
+	//
+	//   ... -> bsvh -> bvm1 -> ... -> bvm#
+	// -------------------------------------------------------------------------
+
+	outs = g.OldestCoinbaseOuts()
+	for i := uint16(0); i < coinbaseMaturity; i++ {
```

Comment:
This bit shouldn't really be required here since a bunch of blocks needed to activate the agenda were added during the advance above and thus the output should already be mature.  I should probably replace it with an assertion that the height is more than the required maturity instead of spending time creating these blocks or only conditionally create them if needed based on some math.

Reply:
Updated to add the aforementioned assertion.